csi/node.go

```diff
@@ -30,3 +52,128 @@ func (s *OsdCsiServer) GetNodeID(ctx context.Context, req *csi.GetNodeIDRequest)
 
 	return result, nil
 }
+
+// NodePublishVolume is a CSI API call which mounts the volume on the specified
+// target path on the node.
+//
+// TODO: Support READ ONLY Mounts
+//
+func (s *OsdCsiServer) NodePublishVolume(
+	ctx context.Context,
+	req *csi.NodePublishVolumeRequest,
+) (*csi.NodePublishVolumeResponse, error) {
+
+	dlog.Debugf("NodePublishVolume req[%#v]", req)
+
+	// Check arguments
+	if req.GetVersion() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Version must be provided")
+	}
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume id must be provided")
+	}
+	if len(req.GetTargetPath()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Target path must be provided")
+	}
+	if req.GetVolumeCapability() == nil || req.GetVolumeCapability().GetAccessMode() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Volume access mode must be provided")
+	}
+
+	// Get volume information
+	v, err := util.VolumeFromName(s.driver, req.GetVolumeId())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Volume id %s not found: %s",
+			req.GetVolumeId(),
+			err.Error())
+	}
+
+	// Gather volume attributes
+	spec, _, _, err := s.specHandler.SpecFromOpts(req.GetVolumeAttributes())
+	if err != nil {
+		return nil, status.Errorf(
+			codes.InvalidArgument,
+			"Invalid volume attributes: %#v",
+			req.GetVolumeAttributes())
+	}
+
+	// This seems weird as a way to change opts to map[string]string
+	opts := make(map[string]string)
+	if len(spec.GetPassphrase()) != 0 {
+		opts[options.OptionsSecret] = spec.GetPassphrase()
+	}
+
+	// Verify target location is an existing directory
+	// See: https://github.com/container-storage-interface/spec/issues/60
+	err = verifyTargetLocation(req.GetTargetPath())
+	if err != nil {
+		return nil, status.Error(codes.NotFound, err.Error())
+	}
+
+	// If this is for a block driver, first attach the volume
+	if s.driver.Type() == api.DriverType_DRIVER_TYPE_BLOCK {
+		_, err := s.driver.Attach(req.GetVolumeId(), opts)
```

Comment:
if _, err := s.driver.Attach(); err != nil {
}

Also at other places in this file.

Reply:
is that a style? afaict, this is ok, right?

Reply:
Yes as mentioned [here](https://github.com/libopenstorage/openstorage/blob/master/STYLEGUIDE.md)

Reply:
Doh!